### PR TITLE
[IOTDB-6103] Adding count_time aggregation feature

### PR DESCRIPTION
**Pull request by Beyyes**
## Description

See user doc: https://apache-iotdb.feishu.cn/docx/SQ11ds2EsokbZ6xZNXDc489HnKc

See design doc: https://apache-iotdb.feishu.cn/docx/Ij0FdZeayoFyvjx1jnrca3MBnkg

### CODE REVIEW

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java [hunk 1]
```diff
@@ -1121,39 +1216,57 @@ private void updateDeviceToSourceTransformAndOutputExpressions(
...
 
       Set<Expression> sourceExpressions = new LinkedHashSet<>();
       for (Expression expression : sourceTransformExpressions) {
-        sourceExpressions.addAll(ExpressionAnalyzer.searchSourceExpressions(expression));
+        sourceExpressions.addAll(searchSourceExpressions(expression));
       }
-      deviceToSourceExpressions.put(deviceName, sourceExpressions);
+      deviceToSourceExpressions
+          .computeIfAbsent(deviceName, key -> new LinkedHashSet<>())
+          .addAll(sourceExpressions);
```
Comment (JackieTien97) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java:1267]: should directly `put` instead of `addAll`

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java [hunk 2]
```diff
@@ -707,6 +779,29 @@ private String analyzeAlias(
...
+      return resultColumnAlias;
+    }
+
+    if (!Objects.equals(normalizedExpression, rawExpression)) {
+      return rawExpression.getOutputSymbol();
+    }
+
+    if (queryStatement.isAggregationQuery()
+        && normalizedExpression.getOutputSymbol().contains(COUNT_TIME)) {
+      return normalizedExpression.getTransformedOutputSymbol();
```
[the remark this answers is not in the document]
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java]: have removed getTransformedOutputSymbol method

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java [hunk 3]
```diff
@@ -649,51 +706,66 @@ private List<Pair<Expression, String>> analyzeSelect(
...
+                ((FunctionExpression) subAggregationExpression).getFunctionName())) {
+
+          for (Expression countTimeSourceExpression :
+              ((FunctionExpression) subAggregationExpression).getCountTimeExpressions()) {
+
+            analyzeExpressionType(analysis, countTimeSourceExpression);
+
+            deviceToSourceTransformExpressions
+                .computeIfAbsent(deviceName, k -> new LinkedHashSet<>())
+                .add(countTimeSourceExpression.getExpressions().get(0));
```
[the remark this answers is not in the document]
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java]: only put countTimeSourceExpression.getExpressions() now

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java [hunk 4]
```diff
@@ -617,16 +667,23 @@ private List<Pair<Expression, String>> analyzeSelect(
...
-              ExpressionAnalyzer.toLowerCaseExpression(measurementExpression);
-          analyzeExpressionType(analysis, normalizedMeasurementExpression);
+          Expression lowerCaseMeasurementExpression = toLowerCaseExpression(measurementExpression);
+          analyzeExpressionType(analysis, lowerCaseMeasurementExpression);
+
+          if (queryStatement.isAggregationQuery()
+              && lowerCaseMeasurementExpression.getOutputSymbol().contains(COUNT_TIME)) {
+            analyzeSelectOfCountTimeAggregationDeviceView(
+                analysis, deviceToSelectExpressionsOfOneMeasurement);
+          }
```
[the remark this answers is not in the document]
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java]: remove this step, all put to `analyzeDeviceToSourceTransform` phase.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java (hunk 1 above)
Comment (Beyyes) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/AnalyzeVisitor.java]: fixed

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/builtin/BuiltinAggregationFunction.java [hunk 5]
```diff
@@ -103,6 +105,7 @@ public static boolean canSplitToMultiPhases(String name) {
       case "mode":
         return true;
       case "count_if":
+      case "count_time":
```
[the remark this answers is not in the document]
Comment (Beyyes) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/udf/builtin/BuiltinAggregationFunction.java:108]: add the judgement in `LogicalPlanVisitor.cannotUseStatistics()`

**Pull request merged**